torchtext/experimental/vocab.py

```diff
@@ -33,30 +22,25 @@ def vocab_from_file_object(file_like_object, **kwargs):
 
     Args:
         file_like_object (FileObject): a file like object to read data from.
-        Remaining keyword arguments: Passed to the constructor of Vocab class.
+        min_freq: The minimum frequency needed to include a token in the vocabulary.
+            Values less than 1 will be set to 1. Default: 1.
+        unk_token: The default unknown token to use. Default: '<unk>'.
+        num_cpus (int): the number of cpus to use when loading the vectors from file. Default: 10.
 
     Returns:
         Vocab: a `Vocab` object.
 
     Examples:
         >>> from torchtext.experimental.vocab import vocab_from_file_object
         >>> f = open('vocab.txt', 'r')
-        >>> v = vocab_from_file_object(f, specials=('<unk>', '<pad>', '<eos>'), specials_first=False)
+        >>> v = vocab_from_file_object(f)
     """
-    ordered_dict = OrderedDict()
-    num_lines = _infer_shape(file_like_object)
-    for line in tqdm(file_like_object, unit_scale=0, unit="lines", total=num_lines):
-        token = line.rstrip()
-        if token in ordered_dict:
-            ordered_dict[token] += 1
-        else:
-            ordered_dict[token] = 1
-
-    return Vocab(ordered_dict, **kwargs)
+    vocab_obj = torch.ops.torchtext._load_vocab_from_file(file_like_object.name, unk_token, min_freq, num_cpus)
+    return Vocab(vocab_obj)
 
 
-class Vocab(nn.Module):
-    r"""Creates a vocab object which maps tokens to indices.
+def vocab(ordered_dict, min_freq=1, unk_token='<unk>'):
```

Comment:
It is counterintuitive to me. We have Vocab class but we don't prefer users to call it directly. In `torchtext.vocab.py`, we have `build_vocab_from_iterator` func. Should we do the similar thing here like `build_vocab`? @cpuhrsch

Reply:
I think this is because we want to follow a similar convention to pytorch `Tensors`. Users are not meant to call the `torch.Tensors` class directly but instead create a tensor object using `torch.tensor()` factory method. This also allows us to directly create the underlying cpp `Vocab` object and pass it into the python `Vocab` class instead of having to pass in params to the `Vocab` class after loading tokens from a file (reduces unnecessary FFI btw cpp->python->cpp->python).

Maybe @cpuhrsch can fill in anything I missed

Reply:
This pattern "cpp->python->cpp->python" only add minor overhead to the performance of the construction time, right? IMO, `vocab` func does exactly same with what `Vocab` constructor did in previous commit.

Reply:
I think the FFI's can actually add significant performance overheads due to the copy constructor being called multiple times. The overhead also increases linearly as the size of the Vocab increases. 

Using this factory pattern would also allows us to decouple the cpp Vocab class from the python Vocab class. In the future if we wanted to choose between 2 different underlying cpp classes during construction, we can do so more easily without having to rewrite our `Vocab `constructor.